FIAT/brezzi_douglas_marini.py

```diff
@@ -73,14 +100,37 @@ def __init__(self, ref_el, degree):
 class BrezziDouglasMarini(finite_element.CiarletElement):
     """The BDM element"""
 
-    def __init__(self, ref_el, degree):
+    def __init__(self, ref_el, degree, variant=None):
+
+        if variant is None:
+            variant = "point"
+            print('Warning: Variant of BDM element will change from point evaluation to integral evaluation.'
+                  'You should project into variant="integral"')
+
+        if not (variant == "point" or "integral" in variant):
+            raise ValueError('Choose either variant="point" or variant="integral"'
+                             'or variant="integral(Quadrature degree)"')
+
+        if variant == "integral":
+            quad_deg = 5 * (degree + 1)
```

Comment:
We don't need it to be this high, but on an edge, it's going to be a lower-order term compared to internal moments.

Reply:
So there are two things here. @FabianL1908 has all the details.

For optimal convergence of the interpolant, we need exact quadrature only against polynomials of degree k (I think).

However, if you want divergence- (or curl-preserving) interpolation of _non-polynomial_ expressions (for example your exact initial condition is div-free, and you want to maintain that, then you need _more_ quadrature: basically we need to integrate well enough that the error is only machine epsilon. Since we can't adapt this quadrature on a per-interpolant basis (the choice defines the primal basis), we are default to something large. The form compiler (which might have a chance of estimating the degree of the expression) can then provide a warning that your interpolant might not be exact.